src/lib/mina_graphql/mina_graphql.ml

```diff
@@ -2198,19 +2198,54 @@ module Queries = struct
         Mina_lib.snark_pool mina |> Network_pool.Snark_pool.resource_pool
         |> Network_pool.Snark_pool.Resource_pool.all_completed_work )
 
+  (* Auxiliary function to fetch all pending work *)
+  let get_pending_work mina =
+    let snark_job_state = Mina_lib.snark_job_state mina in
+    let snark_pool = Mina_lib.snark_pool mina in
+    let fee_opt =
+      Mina_lib.(
+        Option.map (snark_worker_key mina) ~f:(fun _ -> snark_work_fee mina))
+    in
+    Work_selector.pending_work_statements ~snark_pool ~fee_opt snark_job_state
+
   let pending_snark_work =
     field "pendingSnarkWork" ~doc:"List of snark works that are yet to be done"
       ~args:Arg.[]
       ~typ:(non_null @@ list @@ non_null Types.pending_work)
-      ~resolve:(fun { ctx = mina; _ } () ->
-        let snark_job_state = Mina_lib.snark_job_state mina in
-        let snark_pool = Mina_lib.snark_pool mina in
-        let fee_opt =
-          Mina_lib.(
-            Option.map (snark_worker_key mina) ~f:(fun _ -> snark_work_fee mina))
+      ~resolve:(fun { ctx = mina; _ } () -> get_pending_work mina)
+
+  let pending_snark_work_range =
+    field "pendingSnarkWorkRange"
+      ~doc:
+        "Find any sequence of pending snark works between two indexes in the \
+         pending snark work pool"
+      ~args:
+        Arg.
+          [ arg "startingIndex"
+              ~doc:
+                "The first index to be taken from the pending snark work pool"
+              ~typ:(non_null Types.Input.UInt32.arg_typ)
+          ; arg' "endingIndex"
+              ~doc:
+                "The last index to be taken from the pending snark work pool \
+                 (included). If not specified, only the work at index \
+                 [startingIndex] will be returned. An empty list will be \
+                 returned if startingIndex > endingIndex"
+              ~typ:Types.Input.UInt32.arg_typ ~default:Unsigned.UInt32.max_int
+          ]
+      ~typ:(non_null @@ list @@ non_null Types.pending_work)
+      ~resolve:(fun { ctx = mina; _ } () start_idx end_idx ->
+        let start_idx = Unsigned.UInt32.to_int start_idx in
+        (* end_idx is max_int in the default case, in that case we want to take
+           only the element at start_idx. If [start_idx, end_idx] is not a
+           valid range, [] is returned *)
+        let len =
+          if Unsigned.UInt32.(equal end_idx max_int) then 1
+          else if start_idx > Unsigned.UInt32.to_int end_idx then 0
+          else Unsigned.UInt32.to_int end_idx - start_idx + 1
         in
-        Work_selector.pending_work_statements ~snark_pool ~fee_opt
-          snark_job_state )
+        try List.sub ~pos:start_idx ~len @@ get_pending_work mina with _ -> []
```

Comment:
I am fine with a slice here because the Janestreet runtime is O(n), and we won't actually save on ram usage because work selection stores state.available jobs anyway in work_lib.  Since state.availabe jobs is a super set of the pending work, we wouldn't save any required space in the long run.

Reply:
this is also currently a low frequency graphql call